src/bitcoin.cpp

```diff
@@ -0,0 +1,207 @@
+        if (cmd.show_help || cmd.command.empty()) {
+            tfm::format(std::cout, HELP_USAGE, argv[0]);
+        }
+
+        std::vector<std::string> args;
+        if (cmd.command == "gui") {
+            args.emplace_back(cmd.use_multiprocess ? "qt/bitcoin-gui" : "qt/bitcoin-qt");
+        } else if (cmd.command == "daemon") {
+            args.emplace_back(cmd.use_multiprocess ? "bitcoin-node" : "bitcoind");
+        } else if (cmd.command == "rpc") {
+            args.emplace_back("bitcoin-cli");
+            args.emplace_back("-named");
+        } else if (cmd.command == "wallet") {
+            args.emplace_back("bitcoin-wallet");
+        } else if (cmd.command == "tx") {
+            args.emplace_back("bitcoin-tx");
+        } else if (cmd.command == "test") {
+            args.emplace_back("test/test_bitcoin");
+        } else if (cmd.command == "help") {
+            tfm::format(std::cout, HELP_COMMANDS, argv[0]);
+        } else if (cmd.command == "mine") { // undocumented, used by tests
+            args.emplace_back("bitcoin-mine");
+        } else if (cmd.command == "util") { // undocumented, used by tests
+            args.emplace_back("bitcoin-util");
+        } else if (!cmd.command.empty()){
+            throw std::runtime_error(strprintf("Unrecognized command: '%s'", cmd.command));
+        }
+        if (!args.empty()) {
+            args.insert(args.end(), cmd.args.begin(), cmd.args.end());
+            ExecCommand(args, argv[0]);
+        }
+    } catch (const std::exception& e) {
+        tfm::format(std::cerr, "Error: %s\nTry '%s --help' for more information.", e.what(), argv[0]);
+        return EXIT_FAILURE;
+    }
+    return EXIT_SUCCESS;
+}
+
+CommandLine ParseCommandLine(int argc, char* argv[])
+{
+    CommandLine cmd;
+    cmd.args.reserve(argc);
+    for (int i = 1; i < argc; ++i) {
+        std::string_view arg = argv[i];
+        if (!cmd.command.empty()) {
+            cmd.args.emplace_back(arg);
+        } else if (arg == "-m" || arg == "--multiprocess") {
+            cmd.use_multiprocess = true;
+        } else if (arg == "-M" || arg == "--monolithic") {
+            cmd.use_multiprocess = false;
+        } else if (arg == "-v" || arg == "--version") {
+            cmd.show_version = true;
+        } else if (arg == "-h" || arg == "--help") {
+            cmd.show_help = true;
+        } else if (arg.starts_with("-")) {
+            throw std::runtime_error(strprintf("Unknown option: %s", arg));
+        } else if (!arg.empty()) {
+            cmd.command = arg;
+        }
+    }
+    return cmd;
+}
+
+// Execute the specified bitcoind, bitcoin-qt or other command line in `args`
+// using src, bin and libexec directory paths relative to this executable, where
+// the path to this executable is specified `argv0`.
+//
+// This function doesn't currently print anything but can be debugged from
+// the command line using strace like:
+//
+//     strace -e trace=execve -s 10000 build/src/bitcoin ...
```

Comment:
It's fine to leave the comment as is. I couldn't find an equivalent for macOS.

Reply:
The following works on FreeBSD. In case `dtrace` works on macOS:

```sh
dtrace -n 'proc:::exec-success { trace(curpsinfo->pr_psargs); }' -c 'bitcoin -m gui'
```

Maybe how to enable `dtrace` on macOS: https://github.com/suolapeikko/dtrace?tab=readme-ov-file#sample-dtrace-scripts-for-macos

Reply:
```
% dtrace -n 'proc:::exec-success { trace(curpsinfo->pr_psargs); }' -c 'build/src/bitcoin daemon' 
dtrace: system integrity protection is on, some features will not be available
dtrace: failed to initialize dtrace: DTrace requires additional privileges
```

With sudo:

```
dtrace: system integrity protection is on, some features will not be available

dtrace: invalid probe specifier proc:::exec-success { trace(curpsinfo->pr_psargs); }: probe description proc:::exec-success does not match any probes. System Integrity Protection is on
```

I haven't tried disabling System Integrity Protection (SIP) as that link suggests, not necessary a good idea :-) Though the maybe the `csrutil enable --without dtrace` isn't too bad.